src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -0,0 +1,156 @@
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+/**
+ * Worker class to collect insert size metrics.
+ */
+public final class InsertSizeMetricsCollectorSpark implements Serializable{
+    private static final long serialVersionUID = 1L;
+
+    private final JavaRDD<Integer> lengthRDD;
+    private final InsertSizeMetrics metric = new InsertSizeMetrics();
+    final Histogram<Integer> htslibHistogram;
+
+    /**
+     * Constructor and perform map & reduce.
+     * @param filteredReads          reads that pass filters
+     * @param samRgRecords           a list of read groups
+     * @param HistogramMADTolerance  MAD tolerance when producing histogram plot
+     */
+    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
+    public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
+                                           final long pairsCount,
+                                           //final Set<MetricAccumulationLevel> accumLevels,
+                                           final List<SAMReadGroupRecord> samRgRecords,
+                                           final double HistogramMADTolerance) {
+
+        lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
+
+        double lengthArr[] = returnRDDArray(lengthRDD);
+
+        DescriptiveStatistics percent = new DescriptiveStatistics(lengthArr);
+
+        metric.READ_PAIRS          = pairsCount;
+        metric.MIN_INSERT_SIZE     = (int) percent.getMin();
+        metric.MAX_INSERT_SIZE     = (int) percent.getMax();
+
+        metric.WIDTH_OF_10_PERCENT = (int) Math.round(percent.getPercentile(55.0) - percent.getPercentile(45.0));
+        metric.WIDTH_OF_20_PERCENT = (int) Math.round(percent.getPercentile(60.0) - percent.getPercentile(40.0));
+        metric.WIDTH_OF_30_PERCENT = (int) Math.round(percent.getPercentile(65.0) - percent.getPercentile(35.0));
+        metric.WIDTH_OF_40_PERCENT = (int) Math.round(percent.getPercentile(70.0) - percent.getPercentile(30.0));
+        metric.WIDTH_OF_50_PERCENT = (int) Math.round(percent.getPercentile(75.0) - percent.getPercentile(25.0));
+        metric.WIDTH_OF_60_PERCENT = (int) Math.round(percent.getPercentile(80.0) - percent.getPercentile(20.0));
+        metric.WIDTH_OF_70_PERCENT = (int) Math.round(percent.getPercentile(85.0) - percent.getPercentile(15.0));
+        metric.WIDTH_OF_80_PERCENT = (int) Math.round(percent.getPercentile(90.0) - percent.getPercentile(10.0));
+        metric.WIDTH_OF_90_PERCENT = (int) Math.round(percent.getPercentile(95.0) - percent.getPercentile(5.0));
+        metric.WIDTH_OF_99_PERCENT = (int) Math.round(percent.getPercentile(99.5) - percent.getPercentile(0.5));
+
+        metric.MEDIAN_INSERT_SIZE  = percent.getPercentile(50.0);
+
+        final int median = (int) metric.MEDIAN_INSERT_SIZE;
+        final JavaRDD<Integer> absLengthDiffRDD = lengthRDD.map(length -> Math.abs(length - median));
```

Comment:
remember that if you use an RDD twice it will be recomputed again unless you cache it!